### PR TITLE
Add EIP: Minimal Upgradable Proxy Contract

### PR DESCRIPTION
**Pull request by xiaobaiskill**
Discussion at https://ethereum-magicians.org/t/eip-xxxx-minimal-upgradable-proxy/14754

### CODE REVIEW

File: EIPS/eip-7229.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+created: 2023-06-24
+---
+
+## Abstract
+
+This proposal introduces the Minimal Upgradable Contract, a lightweight alternative to the existing upgradable contract implementation provided by OpenZeppelin. The Minimal Upgradable Contract aims to significantly reduce gas consumption during deployment while maintaining upgradability features.
+
+## Motivation
+
+Current upgradable contract solutions, such as OpenZeppelin's [EIP-1967](./eip-1967.md) implementation, often incur high gas costs during deployment. The goal of this proposal is to present an alternative approach that offers a significant reduction in gas consumption while maintaining the ability to upgrade contract logic.
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:18]:
```suggestion
Current upgradable contract solutions, such as OpenZeppelin's [ERC-1967](./eip-1967.md) implementation, often incur high gas costs during deployment. The goal of this proposal is to present an alternative approach that offers a significant reduction in gas consumption while maintaining the ability to upgrade contract logic.
```

File: EIPS/eip-7229.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+3、 Contract Compatibility: The rationale aligns with existing Ethereum Virtual Machine (EVM) environments and does not introduce any compatibility issues or modifications to the underlying infrastructure. It can be easily integrated into the current Ethereum ecosystem, ensuring seamless adoption by developers and users.
+
+4、 Developer-Friendly Implementation: The rationale simplifies the implementation process for upgradeable contracts. By providing a straightforward mechanism to access and delegate calls to the logic contract, developers can focus on contract logic and functionality rather than complex storage management.
+
+By adopting this rationale, we aim to strike a balance between gas efficiency and contract upgradability. It allows developers to deploy upgradeable contracts with minimal gas consumption, making them more accessible and economically viable for a wide range of applications.
+
+
+## Gas Efficiency
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:46]:
```suggestion
### Gas Efficiency
```

File: EIPS/eip-7229.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+Compared to existing upgradable contract solutions, the Minimal Upgradable Contract demonstrates a significant reduction in gas consumption during deployment. While OpenZeppelin's [EIP-1967](./eip-1967.md) implementation may consumes nearly several hundred thousand gas for deployment, the Minimal Upgradable Contract can be deployed with just a few tens of thousands of gas, resulting in substantial cost savings.
+
+- [Transaction deploying the Minimal Upgradable Contract (32bytes slot)](https://goerli.etherscan.io/tx/0x19cf047783a159fd2f44a4dbee9fe0113c0899c29e54744c6f47d062381621cc)
+- [Transaction deploying the Minimal Upgradable Contract (1bytes slot)](https://goerli.etherscan.io/tx/0x96c86fd3a73c00f4a0a35c9837726c059212efb2c60099dc7697d92986c7cc80)
+- [Transaction deploying using OpenZeppelin's EIP-1967](https://goerli.etherscan.io/tx/0xb4677404ac1d5b56a9ad1dd2a34c869738e6fcda07bcc78501354b309ce6e6b1)
+
+## Implementation
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:54]:
```suggestion
### Implementation
```

File: EIPS/eip-7229.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+A reference implementation of the Minimal Upgradable Contract, including the proxy contract and an example implementation contract, will be provided as open-source code. This implementation will serve as a starting point for developers to adopt and customize the Minimal Upgradable Contract in their projects.
+
+### Example implementations
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:58]:
```suggestion
#### Example implementations
```

File: EIPS/eip-7229.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+- [deploy proxy contract when deploying logic contract (32bytes slot)](https://github.com/xiaobaiskill/minimal-upgradable-proxy/blob/main/contracts/mock/mock32/Example.sol)
+- [deploy proxy contract (32bytes slot)](https://github.com/xiaobaiskill/minimal-upgradable-proxy/blob/main/contracts/mock/mock32/DeployContract.sol)
+
+- [deploy proxy contract when deploying logic contract (1bytes slot)](https://github.com/xiaobaiskill/minimal-upgradable-proxy/blob/main/contracts/mock/mock1/Example.sol)
+- [deploy proxy contract (1bytes slot)](https://github.com/xiaobaiskill/minimal-upgradable-proxy/blob/main/contracts/mock/mock1/DeployContract.sol)
+
+### Standard Proxy
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:66]:
```suggestion
#### Standard Proxy
```

File: EIPS/eip-7229.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+PUSH1 0x37          [0x37 result 00 returnDataSize result]
+JUMPI				[00 returnDataSize result]
+REVERT              [result]
+JUMPDEST            [00 returnDataSize result]
+RETURN              [result]
+```
+
+NOTE: To push a zero value onto the stack without abusing the `RETURNDATASIZE` opcode, the above code utilizes [EIP-3855](./eip-3855.md). It achieves this by using the `PUSH0` instruction to push the zero value.
+
+### Storage slot of logic address optimization
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:121]:
```suggestion
#### Storage slot of logic address optimization
```

File: EIPS/eip-7229.md [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+Test cases are available at [https://github.com/xiaobaiskill/minimal-upgradable-proxy/tree/main/test](https://github.com/xiaobaiskill/minimal-upgradable-proxy/tree/main/test)
+
+
+## Security Considerations
+
+None.
+
+## References
+
+- [OpenZeppelin EIP-1967: Standard Proxy Storage Slots](./eip-1967.md)
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:189]:
```suggestion
- [OpenZeppelin ERC-1967: Standard Proxy Storage Slots](./eip-1967.md)
```

File: EIPS/eip-7229.md [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+- [EIP-1822: Universal Upgradeable Proxy Standard (UUPS)](./eip-1822.md)
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:190]:
```suggestion
- [ERC-1822: Universal Upgradeable Proxy Standard (UUPS)](./eip-1822.md)
```

File: EIPS/eip-7229.md [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+- [EIP-1167: Minimal Proxy Contract](./eip-1167.md)
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:191]:
```suggestion
- [ERC-1167: Minimal Proxy Contract](./eip-1167.md)
```

File: EIPS/eip-7229.md [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+
+## Gas Efficiency
+
+Compared to existing upgradable contract solutions, the Minimal Upgradable Contract demonstrates a significant reduction in gas consumption during deployment. While OpenZeppelin's [EIP-1967](./eip-1967.md) implementation may consumes nearly several hundred thousand gas for deployment, the Minimal Upgradable Contract can be deployed with just a few tens of thousands of gas, resulting in substantial cost savings.
+
+- [Transaction deploying the Minimal Upgradable Contract (32bytes slot)](https://goerli.etherscan.io/tx/0x19cf047783a159fd2f44a4dbee9fe0113c0899c29e54744c6f47d062381621cc)
+- [Transaction deploying the Minimal Upgradable Contract (1bytes slot)](https://goerli.etherscan.io/tx/0x96c86fd3a73c00f4a0a35c9837726c059212efb2c60099dc7697d92986c7cc80)
+- [Transaction deploying using OpenZeppelin's EIP-1967](https://goerli.etherscan.io/tx/0xb4677404ac1d5b56a9ad1dd2a34c869738e6fcda07bcc78501354b309ce6e6b1)
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:52]:
```suggestion
- [Transaction deploying using OpenZeppelin's ERC-1967](https://goerli.etherscan.io/tx/0xb4677404ac1d5b56a9ad1dd2a34c869738e6fcda07bcc78501354b309ce6e6b1)
```

File: EIPS/eip-7229.md [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+4、 Developer-Friendly Implementation: The rationale simplifies the implementation process for upgradeable contracts. By providing a straightforward mechanism to access and delegate calls to the logic contract, developers can focus on contract logic and functionality rather than complex storage management.
+
+By adopting this rationale, we aim to strike a balance between gas efficiency and contract upgradability. It allows developers to deploy upgradeable contracts with minimal gas consumption, making them more accessible and economically viable for a wide range of applications.
+
+
+## Gas Efficiency
+
+Compared to existing upgradable contract solutions, the Minimal Upgradable Contract demonstrates a significant reduction in gas consumption during deployment. While OpenZeppelin's [EIP-1967](./eip-1967.md) implementation may consumes nearly several hundred thousand gas for deployment, the Minimal Upgradable Contract can be deployed with just a few tens of thousands of gas, resulting in substantial cost savings.
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:48]:
```suggestion
Compared to existing upgradable contract solutions, the Minimal Upgradable Contract demonstrates a significant reduction in gas consumption during deployment. While OpenZeppelin's [ERC-1967](./eip-1967.md) implementation may consumes nearly several hundred thousand gas for deployment, the Minimal Upgradable Contract can be deployed with just a few tens of thousands of gas, resulting in substantial cost savings.
```

File: EIPS/eip-7229.md [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+
+- [Transaction deploying the Minimal Upgradable Contract (32bytes slot)](https://goerli.etherscan.io/tx/0x19cf047783a159fd2f44a4dbee9fe0113c0899c29e54744c6f47d062381621cc)
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:50]: all links to external sources are banned by the bot, you can maybe add a file with the relevant details in the assets folder corresponding to your eip

File: EIPS/eip-7229.md [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+- [Transaction deploying the Minimal Upgradable Contract (1bytes slot)](https://goerli.etherscan.io/tx/0x96c86fd3a73c00f4a0a35c9837726c059212efb2c60099dc7697d92986c7cc80)
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:51]: same comment about external links

File: EIPS/eip-7229.md (hunk 10 above)
Comment (Joachim-Lebrun) [reply EIPS/eip-7229.md:52]: also same comment on external links here

File: EIPS/eip-7229.md [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+## Test Cases
+
+Test cases are available at [https://github.com/xiaobaiskill/minimal-upgradable-proxy/tree/main/test](https://github.com/xiaobaiskill/minimal-upgradable-proxy/tree/main/test)
+
+
+## Security Considerations
+
+None.
+
+## References
```
Comment (Joachim-Lebrun) [inline EIPS/eip-7229.md:187]: see [EIP-1](https://eips.ethereum.org/EIPS/eip-1) for the name of authorized sections, maybe could be `Reference Implementation` but then it has to come before `Security Considerations`

File: EIPS/eip-7229.md (hunk 12 above)
Comment (xiaobaiskill) [reply EIPS/eip-7229.md]: Thank you for your suggestions. I have made targeted modifications based on your feedback and have added a file containing detailed information related to the proposal. Thank you very much for your contribution.

**Comment by xiaobaiskill** [comment]
all errors are fixed. view please.

File: EIPS/eip-7229.md [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+PUSH0               [00 returnDataSize result] 
+DUP3                [result 00 returnDataSize result]
+PUSH1 0x18          [0x18 result 00 returnDataSize result]
+JUMPI				[00 returnDataSize result]
+REVERT              [result]
+JUMPDEST            [00 returnDataSize result]
+RETURN              [result]
+```
+
+The bytecode generated by the above opcodes is as follows `60xx73yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy8155600960305f3960f81b60095260106039600a39601a5ff3365f5f375f5f365f60545af43d5f5f3e3d5f82601857fd5bf3`, replace `xx` to a slot of 1byte and replace `yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy` to a address of 20bytes before deploying contract
```
Comment (SamWilsn) [inline EIPS/eip-7229.md:177]: You're introducing new content in your rationale section. The rationale section should only explain the stuff in the specification section, and shouldn't introduce any new requirements.

File: EIPS/eip-7229.md [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+PUSH0               [00 returnDataSize result] 
+DUP3                [result 00 returnDataSize result]
+PUSH1 0x17          [0x17 result 00 returnDataSize result]
+JUMPI				[00 returnDataSize result]
+REVERT              [result]
+JUMPDEST            [00 returnDataSize result]
+RETURN              [result]
+```
+
+The bytecode generated by the above opcodes is as follows `73yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy5f55600960285f396010603160093960195ff3365f5f375f5f365f5f545af43d5f5f3e3d5f82601757fd5bf3`, replace `yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy` to a address of 20bytes before deploying contract
```
Comment (SamWilsn) [inline EIPS/eip-7229.md:230]: Same here.

File: EIPS/eip-7229.md [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+
+A reference implementation of the Minimal Upgradable Contract, including the proxy contract and an example implementation contract, will be provided as open-source code. This implementation will serve as a starting point for developers to adopt and customize the Minimal Upgradable Contract in their projects.
+
+#### Example implementations
+
+- [deploy proxy contract when deploying logic contract (32bytes slot)](../assets/eip-7229/contracts/mock/mock32/Example.sol)
+- [deploy proxy contract (32bytes slot)](../assets/eip-7229/contracts/mock/mock32/DeployContract.sol)
+
+- [deploy proxy contract when deploying logic contract (1bytes slot)](../assets/eip-7229/contracts/mock/mock1/Example.sol)
+- [deploy proxy contract (1bytes slot)](../assets/eip-7229/contracts/mock/mock1/DeployContract.sol)
```
Comment (SamWilsn) [inline EIPS/eip-7229.md:65]: Should should move this to the Reference Implementation section (after Test Cases.)

File: EIPS/eip-7229.md [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+
+```
+cd ../assets/eip-7229
+npm install
+npx hardhat test
+```
+
+## Security Considerations
+
+None.
```
Comment (SamWilsn) [inline EIPS/eip-7229.md:245]: I'm sure there are at least some security considerations :upside_down_face: 

```suggestion
<!-- needs discussion -->
```

File: assets/eip-7229/contracts/mock/PayableTokenV1.sol [hunk 19]
```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (SamWilsn) [inline assets/eip-7229/contracts/mock/PayableTokenV1.sol:1]:
```suggestion
// SPDX-License-Identifier: CC0-1.0
```

We can't have UNLICENSED code in the EIPs repository.

File: assets/eip-7229/contracts/mock/PayableTokenV2.sol [hunk 20]
```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (SamWilsn) [inline assets/eip-7229/contracts/mock/PayableTokenV2.sol:1]: No `UNLICENSED`.

File: assets/eip-7229/contracts/mock/SimV1.sol [hunk 21]
```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: MIT
```
Comment (SamWilsn) [inline assets/eip-7229/contracts/mock/SimV1.sol:1]: We prefer `CC0-1.0` but `MIT` is fine too.

File: assets/eip-7229/contracts/mock/mock0/Example.sol [hunk 22]
```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (SamWilsn) [inline assets/eip-7229/contracts/mock/mock0/Example.sol:1]: No `UNLICENSED`.

File: assets/eip-7229/contracts/mock/mock0/PayableToken0V1.sol [hunk 23]
```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (SamWilsn) [inline assets/eip-7229/contracts/mock/mock0/PayableToken0V1.sol:1]: No `UNLICENSED`

File: assets/eip-7229/contracts/mock/mock0/PayableToken0V2.sol [hunk 24]
```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (SamWilsn) [inline assets/eip-7229/contracts/mock/mock0/PayableToken0V2.sol:1]: No `UNLICENSED`.

File: assets/eip-7229/contracts/mock/mock1/Example.sol [hunk 25]
```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (SamWilsn) [inline assets/eip-7229/contracts/mock/mock1/Example.sol:1]: No `UNLICENSED`

File: assets/eip-7229/contracts/mock/mock32/Example.sol [hunk 26]
```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (SamWilsn) [inline assets/eip-7229/contracts/mock/mock32/Example.sol:1]: No `UNLICENSED`.

**Review by SamWilsn — CHANGES_REQUESTED** [review]
You _could_ slim down your example code a bit. We try to keep reference implementations minimal.

---

The big overall comment here is that you aren't proposing a _standard_, you're proposing a specific implementation. We don't generally allow specific implementations to be published as EIPs.

EIPs shine when multiple different parties have to coordinate to implement a system. For example, ERC-20 is a great example of what an EIP should be because every wallet wants to work with every ERC-20 token, and every ERC-20 token wants to work with every wallet.

This EIP doesn't really coordinate anything. If someone wants to use your bytecode, they can just use it without having to coordinate with anyone else.

This isn't me saying your bytecode doesn't have value. It absolutely does, and you should promote it as much as possible. Just not in the EIPs repository.

---

I'm going to close this PR, but feel free to re-open it if you would like to rewrite it as a standard instead of an implementation.

**Pull request closed without merging**